cruise-control/src/main/java/com/linkedin/kafka/cruisecontrol/detector/MaintenanceEventTopicReader.java

```diff
@@ -0,0 +1,356 @@
+        break;
+      default:
+        throw new IllegalStateException(String.format("Unrecognized event type %s", maintenancePlan.maintenanceEventType()));
+    }
+    maintenanceEvents.add(_kafkaCruiseControl.config().getConfiguredInstance(AnomalyDetectorConfig.MAINTENANCE_EVENT_CLASS_CONFIG,
+                                                                             MaintenanceEvent.class,
+                                                                             parameterConfigOverrides));
+  }
+
+  /**
+   * See {@link MaintenanceEventReader#readEvents(Duration)}
+   * This method may block beyond the timeout in order to execute additional logic to create maintenance events from the
+   * maintenance plans retrieved from the relevant topic.
+   *
+   * @param timeout The maximum time to block for retrieving records from the relevant topic.
+   * @return Set of maintenance events, or empty set if none is available after the given timeout expires.
+   */
+  @Override
+  public Set<MaintenanceEvent> readEvents(Duration timeout) throws SamplingException {
+    LOG.debug("Reading maintenance events.");
+    long eventReadPeriodEndMs = _kafkaCruiseControl.timeMs();
+    if (refreshPartitionAssignment()) {
+      _lastEventReadPeriodEndTimeMs = eventReadPeriodEndMs;
+      return Collections.emptySet();
+    }
+
+    long timeoutEndMs = eventReadPeriodEndMs + timeout.toMillis();
+    Set<MaintenanceEvent> maintenanceEvents = new HashSet<>();
+    try {
+      Map<TopicPartition, Long> endOffsets = seekToRelevantOffsets();
+      LOG.debug("Started to consume from maintenance event topic partitions {}.", _currentPartitionAssignment);
+      _consumer.resume(_consumer.paused());
+      Set<TopicPartition> partitionsToPause = new HashSet<>();
+
+      do {
+        ConsumerRecords<String, MaintenancePlan> records = _consumer.poll(timeout);
+        for (ConsumerRecord<String, MaintenancePlan> record : records) {
+          if (record == null) {
+            // This means that the record cannot be parsed because the maintenance event type is not recognized.
+            // It might happen when newer type of maintenance events have been added and the current code is still old.
+            // We simply ignore that metric in this case (see MaintenancePlanSerde#fromBytes).
+            LOG.warn("Cannot parse record, please update your Cruise Control version.");
+            continue;
+          }
+
+          long planGenerationTimeMs = record.value().timeMs();
+          if (planGenerationTimeMs + _maintenancePlanExpirationMs < eventReadPeriodEndMs) {
+            LOG.warn("Discarding the expired plan {}. (Expired: {} Evaluated: {}).", record.value(),
+                     planGenerationTimeMs + _maintenancePlanExpirationMs, eventReadPeriodEndMs);
+          } else if (planGenerationTimeMs >= eventReadPeriodEndMs) {
+            TopicPartition tp = new TopicPartition(record.topic(), record.partition());
```

Comment:
Not all the producers will have the same clock and so time is not monotonically increasing.  Does this still work if that is the case?

Reply:
As (arguably) typically assumed across the participants of Kafka ecosystem, we also assume clocks are _loosely synchronized_. For example, we know that secure communication between brokers and clients can fail due to certificate expiration issues caused by clock skew. Similarly, if a producer has a clock that is significantly behind the others, then its plan can expire "faster" than it expects.
Please also see the comment above, which discusses that the delay can be introduced by Kafka producers:
```
  // A maintenance event has a certain validity period after which it expires and becomes invalid. A delay in handling
  // could be introduced by the Kafka producer that generates the plan, the consumer that retrieves the plan, or the
  // network that connects these clients to the Kafka cluster. Maintenance event topic reader discards the expired events.
```

Reply:
I saw that comment.  My point is that this introduces additional latency and that one broker with a bad clock can block all maintenance events from being processed.

Reply:
> one broker with a bad clock can block all maintenance events from being processed.

I am confused. Unless that broker is the _only_ one that produces maintenance events to the topic, why would all maintenance events be blocked?

Reply:
> one broker with a bad clock can block all maintenance events from being processed.

I am a little confused. Unless that broker is either (1) the only one in the system that generates maintenance events or (2) generates maintenance events every `MaintenanceEventDetector#READ_EVENTS_TIMEOUT`, it will not block the other maintenance events. Am I missing something?